src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,753 @@
+ * "string_failbit" means there is failed operation at std::string related calls.
+ * "posixio_failbit" means a POSIX I/O function failed and errno was set.
+ *
+ * rdstate() function returns stream's current status based on this.
+ * setstate() function only accepts these values as its "bits".
+ */
+enum diostates
+{
+    goodbit = 0x01,
+    eofbit = 0x02,
+    buffer_failbit = 0x04,
+    string_failbit = 0x08,
+    posixio_failbit = 0x10
+};
+
+// Some empty classes for special keywords implementation.
+// Used for accepting endl and flush in ostream::write() functions.
+class endline { };
+class flushbuf { };
+static const endline endl;   // same as std::endl
+static const flushbuf flush; // same as std::flush
+
+// streambuf is just cpbuffer with fixed IOSTREAM_BUFSIZE size
+using streambuf = cpbuffer<IOSTREAM_BUFSIZE>;
+
+// streambuf_base is class which maintains a shread_ptr of streambuf and provides
+// setbuf()/getbuf() functions to getting/setting buffer in output/input streams.
+// It's supposed to be used for providing buffer related interfaces in ostream/istream.
+class streambuf_base
+{
+    protected:
+    std::unique_ptr<streambuf> buf;
+
+    public:
+    // Set current buffer
+    // NOTE: new_buffer can't be nullptr!
+    void setbuf(std::unique_ptr<streambuf> new_buffer) noexcept;
+
+    // Get current buffer
+    std::unique_ptr<streambuf> getbuf() noexcept;
+};
+
+/*
+ * ostream
+ * =-=-=-=
+ *
+ * This class is designed to provide basic functionality around the output. It
+ * has std::unique_ptr streambuf (a cpbuffer with a fixed size which can be changed
+ * in compile-time) and stores file descriptor number in its private space.
+ *
+ * Constructors
+ * ------------
+ *
+ * ostream class has 4 constructor (without counting copy/move constructors):
+ *
+ *      // Simple construction
+ *      dio::ostream output_a;
+ *
+ *      // Construction with moving pre-allocated unique_ptr of buffer
+ *      try {
+ *          std::unique_ptr<streambuf> my_buffer(new streambuf);
+ *      }
+ *      catch(const std::bad_alloc &) {
+ *          printerr("Cannot allocate memory for output buffer!\n");
+ *          return -1;
+ *      }
+ *      dio::ostream output_b(std::move(my_buffer));
+ *
+ *      // Construction with file descriptor No.
+ *      dio::ostream output_c(3);
+ *
+ *      // Construction with combination of fd and moving pre-allocated unique_ptr of buffer
+ *      dio::ostream output_d(3, std::move(my_buffer));
+ *
+ * Also setting (or changing) file descriptor/buffer after construction is available by
+ * setfd()/setbuf() functions.
+ *
+ * copy/move constructors/assignments and destructor
+ * -------------------------------------------------
+ *
+ * ostream copy constructor is prohibited (copy constructor is deleted).
+ * ostream copy assignment is same as copy constructor.
+ *
+ * ostream move constructor is protected. All variable contents except of the buffer
```

Comment:
> I'd like to know: why doesn't the move constructor transfer ownership of the buffer? What is the expected use case such that moving the stream but not transferring the buffer ownership is useful?

It does before but I removed it since there is no purpose for moving objects yet.